swap/htlc.go

```diff
@@ -512,3 +555,220 @@ func (h *HtlcScriptV2) MaxTimeoutWitnessSize() int {
+//	OP_SIZE 32 OP_EQUALVERIFY
+//	OP_HASH160 <ripemd160h(swapHash)> OP_EQUALVERIFY
+//	1
+//	OP_CHECKSEQUENCEVERIFY
+func GenClaimPathScript(
+	receiverHtlcKey [32]byte, swapHash lntypes.Hash) ([]byte, error) {
+	builder := txscript.NewScriptBuilder()
+
+	builder.AddData(receiverHtlcKey[:])
+	builder.AddOp(txscript.OP_CHECKSIGVERIFY)
+	builder.AddOp(txscript.OP_SIZE)
+	builder.AddInt64(32)
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+	builder.AddOp(txscript.OP_HASH160)
+	builder.AddData(input.Ripemd160H(swapHash[:]))
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+	builder.AddInt64(1)
+	builder.AddOp(txscript.OP_CHECKSEQUENCEVERIFY)
+	return builder.Script()
+}
+
+// genControlBlock constructs the control block with the specified.
+func (h *HtlcScriptV3) genControlBlock(leafScript []byte) ([]byte, error) {
+	var outputKeyYIsOdd bool
+
+	if h.TaprootKey.SerializeCompressed()[0] == secp.PubKeyFormatCompressedOdd {
+		outputKeyYIsOdd = true
+	}
+
+	leaf := txscript.NewBaseTapLeaf(leafScript)
+	proof := leaf.TapHash()
+
+	controlBlock := txscript.ControlBlock{
+		InternalKey:     h.InternalPubKey,
+		OutputKeyYIsOdd: outputKeyYIsOdd,
+		LeafVersion:     txscript.BaseLeafVersion,
+		InclusionProof:  proof[:],
+	}
+	controlBlockBytes, err := controlBlock.ToBytes()
+
+	if err != nil {
+		return nil, err
+	}
+	return controlBlockBytes, nil
+}
+
+// genSuccessWitness returns the success script to spend this htlc with
+// the preimage.
+func (h *HtlcScriptV3) genSuccessWitness(
+	receiverSig []byte, preimage lntypes.Preimage) (wire.TxWitness, error) {
+
+	controlBlockBytes, err := h.genControlBlock(h.TimeoutScript)
+	if err != nil {
+		return nil, err
+	}
+	return wire.TxWitness{
+		preimage[:],
+		receiverSig,
+		h.ClaimScript,
+		controlBlockBytes,
+	}, nil
+}
+
+// GenTimeoutWitness returns the timeout script to spend this htlc after
+// timeout.
+func (h *HtlcScriptV3) GenTimeoutWitness(
+	senderSig []byte) (wire.TxWitness, error) {
+
+	controlBlockBytes, err := h.genControlBlock(h.ClaimScript)
+	if err != nil {
+		return nil, err
+	}
+	return wire.TxWitness{
+		senderSig,
+		h.TimeoutScript,
+		controlBlockBytes,
+	}, nil
+}
+
+// IsSuccessWitness checks whether the given stack is valid for
+// redeeming the htlc.
+func (h *HtlcScriptV3) IsSuccessWitness(witness wire.TxWitness) bool {
+	return len(witness) == 4
+}
+
+// Script is not implemented, but necessary to conform to interface.
+func (h *HtlcScriptV3) Script() []byte {
+	return nil
+}
+
+// MaxSuccessWitnessSize returns the maximum witness size for the
+// success case witness.
+func (h *HtlcScriptV3) MaxSuccessWitnessSize() int {
+	// Calculate maximum success witness size
+	//
+	// - number_of_witness_elements: 1 byte
+	// - sig_length: 1 byte
+	// - sig: 73 bytes
+	// - preimage_length: 1 byte
+	// - preimage: 32 bytes
+	// - witness_script_length: 1 byte
+	// - witness_script: len(script) bytes
+	// - control_block_length: 1 byte
+	// - control_block: 4129 bytes
```

Comment:
I _think_ this should be based on the worst-case size of our control block so that we can get as close to accurate as possible.

Reply:
I agree with Carla, since our taptree is shallow, our control block is also small compared to the worst theoretical case, so we should calculate the max for our case imo. (Even though we don't use this function).